### PR TITLE
[SPARK-8738] [SQL] [PYSPARK] capture SQL AnalysisException in Python API

### PR DESCRIPTION
**Pull request by davies**
Capture the AnalysisException in SQL, hide the long java stack trace, only show the error message.

cc @rxin

### CODE REVIEW

File: python/pyspark/sql/utils.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import py4j
+
+
+class AnalysisException(Exception):
```
Comment (rxin) [inline python/pyspark/sql/utils.py]: can you add more inline comments explaining how this works?

File: python/pyspark/sql/utils.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Fail to analysis SQL query plan.
```
Comment (rxin) [inline python/pyspark/sql/utils.py]: "Failed to analyze a SQL query plan."

File: python/pyspark/sql/context.py [hunk 3]
```diff
@@ -96,6 +97,7 @@ def __init__(self, sparkContext, sqlContext=None):
         self._jvm = self._sc._jvm
         self._scala_SQLContext = sqlContext
         _monkey_patch_RDD(self)
+        install_exception_handler()
```
Comment (rxin) [inline python/pyspark/sql/context.py]: is this called for hive context as well?

File: python/pyspark/sql/utils.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Hook an exception handler into Py4j, which could capture some SQL exceptions in Java.
+
+    When calling Java API, it will call `get_return_value` to parse the returned object.
+    If any exception happened in JVM, the result will be Java exception object, it raise
+    py4j.protocol.Py4JJavaError. We replace the original `get_return_value` with one that
+    could capture the Java exception and throw a Python one (with the same error message).
+    """
+    old_func = py4j.protocol.get_return_value
+    new_func = capture_sql_exception(old_func)
```
Comment (rxin) [inline python/pyspark/sql/utils.py]: given it is possible to create multiple sql contexts, i think you want to make this idempotent. maybe have a global flag to make sure we don't wrap this too many times.

Comment (davies) [inline python/pyspark/sql/utils.py]: It's already idempotent, the patched function only used inside py4j.java_gateway, the original still in py4j.protocol.

File: python/pyspark/sql/context.py (hunk 3 above)
Comment (davies) [inline python/pyspark/sql/context.py]: yes

File: python/pyspark/sql/utils.py (hunk 4 above)
Comment (rxin) [inline python/pyspark/sql/utils.py]: can you add this as inline comment?

**Comment by rxin** [comment]
lgtm otherwise. we should consider adding this for parser exception in the future also.

**Comment by davies** [comment]
Right now, we throw RuntimeException while parsing.

**Pull request closed without merging**